### PR TITLE
REST API: add functionality for saving Business Address widget in JPO (alternative)

### PR DESCRIPTION
**Pull request by AnnaMag**
An alternative approach to #8480  and #8426. 

#### Changes proposed in this Pull Request:
* adds functions required to inject a Business Address widget or update an existing one,
* saves Business Address data in an option,
* enables widget module if inactive ( throws an error on failure ).

#### Testing instructions:

* Follow the instructions https://github.com/Automattic/jetpack/pull/8426 tested on a site with a Business Plan ( or temporarily set `isBusiness` to `true` in main.jsx). 
* To test that it works for the initially inactive `widgets` module: 
 use `wp jetpack module deactivate` to deactivate the module and follow the previous instructions. After, verify that the module is listed as active in `wp jetpack module list` and that the widget has been set correctly. 
* Verify that the `jpo_business_address` option is set i.e. using `wp option get jpo_business_address`.

### CODE REVIEW

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 1]
```diff
@@ -1053,6 +1053,10 @@ private function _process_onboarding( $data ) {
 			}
 		}
 
+		if ( isset( $data['businessAddress'] ) ) {
+			self::add_business_address( $data );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We still need to handle the errors in the same way that we handle them for the rest of the onboarding settings (adding them to the `error` array). So you might want to update `add_business_address()` to return either an array of errors, a `WP_Error` object, or `true` on success; and if there are any errors, merge them with `$errors`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 2]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
 			: join( ', ', $error );
 	}
 
+	static function get_first_sidebar() {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: How about just moving this one the widgets lib and using it statically from there?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 3]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+	/**
+	 * Insert a new widget in a given sidebar.
+	 *
+	 * @param string $widget_id ID of the widget.
+	 *
+	 * @param array  $widget_options Content of the widget.
+ 	 *
+ 	 * @param string  $sidebar ID of the sidebar to which the widget will be added.
+	 */
+	static function insert_widget_in_sidebar( $widget_id, $widget_options, $sidebar ) {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: How about just moving this one the widgets lib and using it statically from there?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 4]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+	/**
+	 * Update the content of an exisitng widget in a given sidebar.
+	 *
+	 * @param string $widget_id ID of the widget.
+	 *
+	 * @param array  $widget_options New content for the update.
+ 	 *
+ 	 * @param string  $sidebar ID of the sidebar to which the widget will be added.
+	 */
+	static function update_widget_in_sidebar( $widget_id, $widget_options, $sidebar ) {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: How about just moving this one the widgets lib and using it statically from there?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 5]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+				'showmap' => false,
+				'email' => ''
+			);
+
+			if ( ! self::has_business_address_widget( $first_sidebar ) ) {
+				self::insert_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			} else {
+				self::update_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			}
+			wp_send_json_success( array( 'updated' => true ) );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We're not supposed to call `wp_send_json_success` here, because it'll end the request in an unexpected way. `Jetpack_Core_API_Data::update_data()` already handles this by either calling a `rest_ensure_response()` on success, or returning a `WP_Error` object on failure, so we only need to pass on the errors, if any, and success messages, if any (following the pattern that `_process_onboarding` already implements).

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 6]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+			die();
+		}
+		wp_send_json_success( array( 'updated' => false ) );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We're not supposed to call `wp_send_json_success` here, because it'll end the request in an unexpected way. `Jetpack_Core_API_Data::update_data()` already handles this by either calling a `rest_ensure_response()` on success, or returning a `WP_Error` object on failure, so we only need to pass on the errors, if any, and success messages, if any (following the pattern that `_process_onboarding` already implements).

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 7]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+				'email' => ''
+			);
+
+			if ( ! self::has_business_address_widget( $first_sidebar ) ) {
+				self::insert_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			} else {
+				self::update_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			}
+			wp_send_json_success( array( 'updated' => true ) );
+			die();
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We shouldn't prematurely end the request, so the `end()` can be removed.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 8]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+		$widget_instances = get_option( 'widget_' . $widget_id, array() );
+		// Retrieve the key of the next widget instance
+		$numeric_keys = array_filter( array_keys( $widget_instances ), 'is_int' );
+		$next_key = $numeric_keys ? max( $numeric_keys ) + 1 : 2;
+		// Add this widget to the sidebar
+		if ( ! isset( $sidebars_widgets[ $sidebar ] ) ) {
+			$sidebars_widgets[ $sidebar ] = array();
+		}
+		$sidebars_widgets[ $sidebar ][] = $widget_id . '-' . $next_key;
+		// Add the new widget instance
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Nit: Adding a newline before each of those `//` comments, will make the methods you introduce more readable.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 9]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+		return array_shift( $active_sidebars_keys );
+	}
+
+	static function add_business_address( $data ) {
+		$first_sidebar = self::get_first_sidebar();
+
+		if ( $first_sidebar ) {
+			$title = $data['businessAddress'][ 'name' ];
+			$address =
+					$data[ 'businessAddress' ][ 'city' ] . ' ' .
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Indentation appears a little off here. Do we want to make this one a little smarter, by grabbing all values, then filtering out any empty ones, and then building the string?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 10]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+		}
+		$active_sidebars_keys = array_keys( $active_sidebars );
+		return array_shift( $active_sidebars_keys );
+	}
+
+	static function add_business_address( $data ) {
+		$first_sidebar = self::get_first_sidebar();
+
+		if ( $first_sidebar ) {
+			$title = $data['businessAddress'][ 'name' ];
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Code style: we don't wrap specific array keys with leading/trailing spaces, so instances like `[ 'name' ]` need to become like this: `['name']`.

Contrary to this, variable keys need to have trailing/leading spaces, like this: `[ $key ]`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 11]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+			$widget_options = array(
+				'title'   => $title,
+				'address' => $address,
+				'phone'   => '',
+				'hours'   => '',
+				'showmap' => false,
+				'email' => ''
+			);
+
+			if ( ! self::has_business_address_widget( $first_sidebar ) ) {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Do we also want in this PR to store the business address in an option to make it available for the REST API? Should be a quick one.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 12]
```diff
@@ -1066,6 +1070,126 @@ private function _process_onboarding( $data ) {
...
+				self::insert_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			} else {
+				self::update_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Let's make sure to capture errors from those methods and pass them on to `$errors`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 11 above)
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We need to make sure the widgets module is active, you can borrow the code from here: 

https://github.com/Automattic/jetpack/pull/8481/files#diff-2ccec9069213650359171727de9a2471R1048

Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We already have that in the pipeline, I believe https://github.com/Automattic/jetpack/pull/8481.

Comment (tyxla) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: How do you feel about handling these in the same PR? It's only those 4 lines.

Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Yes, we can 👍 closing it then.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 9 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Where are you prettify? 😆

**Comment by AnnaMag** [comment]
Thanks for reviewing @tyxla. Nice sugestions!

The patch has been adapted 👍 I'd appreciate another look when you get a chance. Thanks!

**Comment by AnnaMag** [comment]
One thing left to decide is the order of field in the address: street, city, zip, state. It is a bit tricky as conventions differ per geo-locations. 
Also, the entire address is pasted in a single line, which if long, will look off. Should we add a new line somewhere? i.e. after street and city? Please test a patch with a long address to get a feeling of how that would end up looking.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 13]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
 			: join( ', ', $error );
 	}
 
+	/**
+	 * Retrieve the first active sidebar.
+	 *
+	 * @return array|WP_Error First active sidebar, error if none exists.
+	*/
+	static function get_first_sidebar() {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: How about moving this one to the `widgets` lib too? It's generic and not directly connected to the business address logic.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 14]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+		$widgets_module_active = Jetpack::is_module_active( 'widgets' );
+		if ( ! $widgets_module_active ) {
+			$widgets_module_active = Jetpack::activate_module( 'widgets', false, false );
+		}
+		if ( ! $widgets_module_active ) {
+			return new WP_Error( 'invalid_data', 'Failed to activate module.', 400 );
+		}
+
+		if ( $first_sidebar ) {
+			$title = $business_address_content['name'];
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We need to sanitize this one, because right now we'll insert them plain, and that will allow for scripts and other malicious contents. For sanitization you can use `sanitize_text_field()`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 15]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+			$address =
+				$business_address_content['city'] . ' ' .
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We need to sanitize all those fields, because right now we'll insert them plain, and that will allow for scripts and other malicious contents. For sanitization you can use `sanitize_text_field()`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 16]
```diff
@@ -1053,6 +1054,14 @@ private function _process_onboarding( $data ) {
 			}
 		}
 
+		if ( isset( $data['businessAddress'] ) ) {
+			$handled_business_address = self::handle_business_address( $data['businessAddress'] );
+			if ( is_wp_error( $handled_business_address ) ) {
+				$error[] = 'BusinessAddress';
+			}
+			update_option( 'jpo_business_address', $data['businessAddress'] );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Right now we update the business address regardless of whether it worked properly or not. We should move this to the bottom `handle_business_address()`, right after the insert/update operation has completed successfully.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 17]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+	 * Retrieve the first active sidebar.
+	 *
+	 * @return array|WP_Error First active sidebar, error if none exists.
+	*/
+	static function get_first_sidebar() {
+		$active_sidebars = get_option( 'sidebars_widgets', array() );
+		unset( $active_sidebars[ 'wp_inactive_widgets' ], $active_sidebars[ 'array_version' ] );
+
+		if ( empty( $active_sidebars ) ) {
+			return new WP_Error( 'invalid_data', 'Invalid data.', 400 );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: I think this is not a case when we want to return a `WP_Error`; this is a case when we want to fail silently because there are no sidebars registered. Returning `false` should probably make more sense, especially because down there have `if ( $first_sidebar ) {` when inserting/updating the widget.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 18]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+		}
+		$active_sidebars_keys = array_keys( $active_sidebars );
+		return array_shift( $active_sidebars_keys );
+	}
+
+	/**
+	 * Add or update Business Address widget.
+	 *
+	 * @param array $business_address_content Value corresponding to the Business
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Too long param description, can we shorten? `Array of business address fields` sounds descriptive enough. We should keep those on a single line.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 19]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+	 * Address as received from the settings endpoint.
+	 *
+	 * @return WP_Error|true True when data has been saved correctly,
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We should keep those docs on a single line.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 20]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+	*/
+	static function handle_business_address( $business_address_content ) {
+		$first_sidebar = self::get_first_sidebar();
+
+		$widgets_module_active = Jetpack::is_module_active( 'widgets' );
+		if ( ! $widgets_module_active ) {
+			$widgets_module_active = Jetpack::activate_module( 'widgets', false, false );
+		}
+		if ( ! $widgets_module_active ) {
+			return new WP_Error( 'invalid_data', 'Failed to activate module.', 400 );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: This is not really a `invalid_data` error, I'd expect seeing something like `module_activation_failed` or something.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 21]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+	/**
+	 * Add or update Business Address widget.
+	 *
+	 * @param array $business_address_content Value corresponding to the Business
+	 * Address as received from the settings endpoint.
+	 *
+	 * @return WP_Error|true True when data has been saved correctly,
+	 * error otherwise.
+	*/
+	static function handle_business_address( $business_address_content ) {
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Code would be way shorter if we rename `business_address_content` to just `address`.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 22]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+		}
+		if ( ! $widgets_module_active ) {
+			return new WP_Error( 'invalid_data', 'Failed to activate module.', 400 );
+		}
+
+		if ( $first_sidebar ) {
+			$title = $business_address_content['name'];
+			$address =
+				$business_address_content['city'] . ' ' .
+				$business_address_content['state'] . ' ' .
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: A small suggestion here: since we need to sanitize these and clean out any empty ones, how about something like this (I know it might look a little repetitive, but it's explicit enough):

```
$title = isset( $address['name'] ) ? sanitize_text_field( $address['name'] ) : '';

$street = isset( $address['street'] ) ? sanitize_text_field( $address['street'] ) : '';
$city = isset( $address['city'] ) ? sanitize_text_field( $address['city'] ) : '';
$state = isset( $address['state'] ) ? sanitize_text_field( $address['state'] ) : '';
$zip = isset( $address['zip'] ) ? sanitize_text_field( $address['zip'] ) : '';

$full_address = implode( ' ', array_filter( array( $street, $city, $state, $zip ) ) );
```

Also, upon success, you can save `jpo_business_address`, like this (assuming you remove the other `update_option( 'jpo_business_address'` call):

```
$address_pieces = array(
	'name' => $title,
	'street' => $street,
	'city' => $city,
	'state' => $state,
	'zip' => $zip
);
update_option( 'jpo_business_address', $address_pieces );
```

It might look a little verbose, but it's essential that we take advantage of sanitization before saving both in the widget and in the option.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 23]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+
+			$widget_inserted = '';
+			$widget_updated = '';
+			if ( ! self::has_business_address_widget( $first_sidebar ) ) {
+				$widget_inserted = Jetpack_Widgets::insert_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			} else {
+				$widget_updated = Jetpack_Widgets::update_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			}
+			if ( is_wp_error( $widget_inserted ) || is_wp_error( $widget_updated ) ) {
+				return new WP_Error( 'invalid_data', 'Invalid update.', 400 );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Could we be a little more specific about this error type and error message?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 24]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+			}
+
+			return true;
+		}
+		// No sidebar to place the widget
+		return new WP_Error( 'invalid_data', 'Invalid data.', 400 );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Could we be a little more specific about this error type and error message?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 25]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+	}
+
+	/**
+	 * Check if a given sidebar already contains a widget with a Business Address.
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: We don't really check whether we have a Business Address; rather we check whether there's a Jetpack "Contact Info & Map" (`widget_contact_info`) widget in the given sidebar.

File: _inc/lib/widgets.php [hunk 26]
```diff
@@ -676,4 +676,86 @@ public static function get_registered_widget_object( $id_base ) {
...
+		if ( ! isset( $sidebars_widgets[ $sidebar ] ) ) {
+			$sidebars_widgets[ $sidebar ] = array();
+		}
+		$sidebars_widgets[ $sidebar ][] = $widget_id . '-' . $next_key;
+
+		// Add the new widget instance
+		$widget_instances[ $next_key ] = $widget_options;
+
+		// Store updated sidebars, widgets and their instances
+		if ( ! ( update_option( 'blogname', $data['siteTitle'] ) || get_option( 'blogname' ) == $data['siteTitle'] )
```
Comment (tyxla) [inline _inc/lib/widgets.php]: This line seems off and unnecessary - we even don't have a `$data` here. This will probably throw an error when inserting a sidebar (although insertion will be successful).

File: _inc/lib/widgets.php [hunk 27]
```diff
@@ -676,4 +676,86 @@ public static function get_registered_widget_object( $id_base ) {
...
+		|| ( ! ( update_option( 'sidebars_widgets', $sidebars_widgets ) ) )
+		|| ( ! ( update_option( 'widget_' . $widget_id, $widget_instances ) ) ) ) {
+			return new WP_Error( 'invalid_data', 'Invalid update.', 400 );
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Could we be a little more specific about this error type and error message?

File: _inc/lib/widgets.php [hunk 28]
```diff
@@ -676,4 +676,86 @@ public static function get_registered_widget_object( $id_base ) {
...
+		if ( ! $widget_key ) {
+			return new WP_Error( 'invalid_data', 'No such widget.', 400 );
+		}
+
+		// Update the widget instance with the new data
+		$widget_instances[ $widget_key ] = array_merge( $widget_instances[ $widget_key ], $widget_options );
+
+		// Store updated widget instances and return Error when not successful
+		if ( ! ( update_option( 'widget_' . $widget_id, $widget_instances ) ) ) {
+			return new WP_Error( 'invalid_data', 'Invalid update.', 400 );
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Could we be a little more specific about this error type and error message?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 29]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
 			: join( ', ', $error );
 	}
 
+	/**
+	 * Retrieve the first active sidebar.
+	 *
+	 * @return array|WP_Error First active sidebar, error if none exists.
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Should `array` actually be `string` here?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 30]
```diff
@@ -1066,6 +1075,95 @@ private function _process_onboarding( $data ) {
...
+			if ( ! self::has_business_address_widget( $first_sidebar ) ) {
+				$widget_inserted = Jetpack_Widgets::insert_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			} else {
+				$widget_updated = Jetpack_Widgets::update_widget_in_sidebar( 'widget_contact_info', $widget_options, $first_sidebar );
+			}
+			if ( is_wp_error( $widget_inserted ) || is_wp_error( $widget_updated ) ) {
+				return new WP_Error( 'invalid_data', 'Invalid update.', 400 );
+			}
+
+			return true;
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Updating the `jpo_business_address` option should probably happen above this line (right after we've made sure updating/saving has succeeded).

File: _inc/lib/widgets.php [hunk 31]
```diff
@@ -676,4 +676,86 @@ public static function get_registered_widget_object( $id_base ) {
 	public static function validate_id_base( $id_base ) {
 		return ( false !== self::get_registered_widget_object( $id_base ) );
 	}
+
+	/**
+	 * Insert a new widget in a given sidebar.
+	 *
+	 * @param string $widget_id ID of the widget.
+	 *
```
Comment (tyxla) [inline _inc/lib/widgets.php]: No need for empty lines between `@param`s.

File: _inc/lib/widgets.php [hunk 32]
```diff
@@ -676,4 +676,86 @@ public static function get_registered_widget_object( $id_base ) {
...
+		};
+
+		return true;
+	}
+
+	/**
+	 * Update the content of an exisitng widget in a given sidebar.
+	 *
+	 * @param string $widget_id ID of the widget.
+	 *
```
Comment (tyxla) [inline _inc/lib/widgets.php]: No need for empty lines between `@param`s.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 17 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Updated to `false`. 
Side note: from a user perspective it does not feel like a good UX to not inform them why the business address will not appear on their site despite having completed the step. Should we possibly enable a sidebar on their behalf?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 13 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: yes, missed that one 👍

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 22 above)
Comment (AnnaMag) [reply _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Yes, previous version was missing filtering empty info and sanitization. Good points 👍

**Comment by AnnaMag** [comment]
Feedback incorporated 👍 Thanks @tyxla!

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 33]
```diff
@@ -1066,6 +1074,89 @@ private function _process_onboarding( $data ) {
...
+	*/
+	static function handle_business_address( $address ) {
+		$first_sidebar = Jetpack_Widgets::get_first_sidebar();
+
+		$widgets_module_active = Jetpack::is_module_active( 'widgets' );
+		if ( ! $widgets_module_active ) {
+			$widgets_module_active = Jetpack::activate_module( 'widgets', false, false );
+		}
+		if ( ! $widgets_module_active ) {
+			return new WP_Error( 'module_activation_failed', 'Failed to activate module.', 400 );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Maybe worth clarifying which module we failed to activate: `Failed to activate the widgets module.`

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 34]
```diff
@@ -1066,6 +1074,89 @@ private function _process_onboarding( $data ) {
...
+		}
+
+		if ( $first_sidebar ) {
+			$title = isset( $address['name'] ) ? sanitize_text_field( $address['name'] ) : '';
+			$street = isset( $address['street'] ) ? sanitize_text_field( $address['street'] )
+			: '';
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Let's keep this on the same line.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 35]
```diff
@@ -1066,6 +1074,89 @@ private function _process_onboarding( $data ) {
...
+				'title'   => $title,
+				'address' => $full_address,
+				'phone'   => '',
+				'hours'   => '',
+				'showmap' => false,
+				'email' => ''
+			);
+
+			$widget_inserted = '';
+			$widget_updated = '';
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Why do we need 2 separate variables for handling insert and update if we return the same error message?

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 36]
```diff
@@ -1066,6 +1074,89 @@ private function _process_onboarding( $data ) {
...
+				'city' => $city,
+				'state' => $state,
+				'zip' => $zip
+			);
+			update_option( 'jpo_business_address', $address_save );
+			return true;
+		}
+
+		// No sidebar to place the widget
+		return new WP_Error( 'sidebar_failed', 'No sidebar.', 400 );
```
Comment (tyxla) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: I'd rephrase `sidebar_failed` to `sidebar_not_found`.

File: _inc/lib/widgets.php [hunk 37]
```diff
@@ -676,4 +676,96 @@ public static function get_registered_widget_object( $id_base ) {
...
+		if ( ! ( update_option( 'sidebars_widgets', $sidebars_widgets ) )
+		|| ( ! ( update_option( 'widget_' . $widget_id, $widget_instances ) ) ) )  {
+			return new WP_Error( 'widget_update_failed', 'Failed to update widget or sidebar.', 400 );
+		};
+
+		return true;
+	}
+
+	/**
+	 * Update the content of an exisitng widget in a given sidebar.
```
Comment (tyxla) [inline _inc/lib/widgets.php]: We've got a typo here: `exisitng`

File: _inc/lib/widgets.php [hunk 38]
```diff
@@ -676,4 +676,96 @@ public static function get_registered_widget_object( $id_base ) {
...
+	 *
+	 * @param string $widget_id ID of the widget.
+	 * @param array $widget_options New content for the update.
+ 	 * @param string $sidebar ID of the sidebar to which the widget will be added.
+ 	 *
+ 	 * @return WP_Error|true True when data has been updated correctly,
+ 	 * error otherwise.
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Let's keep this on the same line.

File: _inc/lib/widgets.php [hunk 39]
```diff
@@ -676,4 +676,96 @@ public static function get_registered_widget_object( $id_base ) {
...
+
+	/**
+	 * Insert a new widget in a given sidebar.
+	 *
+	 * @param string $widget_id ID of the widget.
+	 * @param array $widget_options Content of the widget.
+ 	 * @param string $sidebar ID of the sidebar to which the widget will be added.
+ 	 *
+ 	 * @return WP_Error|true True when data has been saved correctly,
+ 	 * error otherwise.
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Let's keep this on the same line.

File: _inc/lib/widgets.php [hunk 40]
```diff
@@ -676,4 +676,96 @@ public static function get_registered_widget_object( $id_base ) {
...
+	static function update_widget_in_sidebar( $widget_id, $widget_options, $sidebar ) {
+		// Retrieve sidebars, widgets and their instances
+		$sidebars_widgets = get_option( 'sidebars_widgets', array() );
+		$widget_instances = get_option( 'widget_' . $widget_id, array() );
+
+		// Retrieve index of first widget instance in that sidebar
+		$widget_key = false;
+		foreach ( $sidebars_widgets[ $sidebar ] as $widget ) {
+			if ( strpos( $widget, 'widget_contact_info' ) !== false ) {
+				$widget_key = absint( str_replace( 'widget_contact_info-', '', $widget ) );
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Should we use `$widget_id` instead of `'widget_contact_info'` here?

**Comment by tyxla** [comment]
@dereksmart @zinigor can we please get a review of this one? Thank you!

File: _inc/lib/widgets.php (hunk 40 above)
Comment (AnnaMag) [reply _inc/lib/widgets.php]: yes!

**Comment by AnnaMag** [comment]
All points addressed 👍 . Thanks for an awesome review @tyxla !

File: _inc/lib/widgets.php [hunk 41]
```diff
@@ -715,14 +714,13 @@ static function insert_widget_in_sidebar( $widget_id, $widget_options, $sidebar
 	}
 
 	/**
-	 * Update the content of an exisitng widget in a given sidebar.
+	 * Update the content of an exisiting widget in a given sidebar.
```
Comment (tyxla) [inline _inc/lib/widgets.php]: Should be `existing` instead of `exisiting`

Comment (AnnaMag) [reply _inc/lib/widgets.php]: 🤦‍♀️

Comment (AnnaMag) [reply _inc/lib/widgets.php]: s-e-r-i-o-u-s-l-y

File: _inc/lib/widgets.php [hunk 42]
```diff
@@ -676,4 +676,94 @@ public static function get_registered_widget_object( $id_base ) {
...
+			$sidebars_widgets[ $sidebar ] = array();
+		}
+		$sidebars_widgets[ $sidebar ][] = $widget_id . '-' . $next_key;
+
+		// Add the new widget instance
+		$widget_instances[ $next_key ] = $widget_options;
+
+		// Store updated sidebars, widgets and their instances
+		if ( ! ( update_option( 'sidebars_widgets', $sidebars_widgets ) )
+		|| ( ! ( update_option( 'widget_' . $widget_id, $widget_instances ) ) ) )  {
```
Comment (zinigor) [inline _inc/lib/widgets.php]: Minor nitpick: can we please have the `if` indented this way for readability?
```
if (
    condition
    || condition2
) {
```

**Comment by AnnaMag** [comment]
Thanks for the review @zinigor! Styling addressed 👍 
Could you have another look, please? Feel free to merge :)

**Comment by zinigor** [comment]
Thanks! Please merge once Travis passes.
